src/main/scala/is/hail/keytable/KeyTable.scala

```diff
@@ -62,44 +32,40 @@ object KeyTable {
   }
 }
 
-case class KeyTable(@transient hc: HailContext, @transient rdd: RDD[(Annotation, Annotation)],
-  keySignature: TStruct, valueSignature: TStruct) extends Serializable {
-  require(fieldNames.areDistinct())
-
-  private val localValueSignature = valueSignature
-  val (signature, mergeKeyAndValue) = keySignature.merge(localValueSignature)
+case class KeyTable(@transient hc: HailContext, @transient rdd: RDD[Annotation],
+  signature: TStruct, keyNames: Array[String]) {
+  require(fieldNames.areDistinct() && keyNames.areDistinct())
+  require(keyNames.forall(fieldNames.contains(_)))
 
   def fields = signature.fields
 
-  def keySchema = keySignature.schema
-
-  def valueSchema = valueSignature.schema
+  def keyFields = keyNames.flatMap { n => signature.fieldOption(n) }
 
   def schema = signature.schema
 
-  def keyNames = keySignature.fields.map(_.name).toArray
-
-  def valueNames = valueSignature.fields.map(_.name).toArray
-
-  def fieldNames = keyNames ++ valueNames
+  def fieldNames = fields.map(_.name).toArray
 
   def nRows = rdd.count()
 
   def nFields = fields.length
 
-  def nKeys = keySignature.size
-
-  def nValues = valueSignature.size
+  def nKeys = keyNames.length
 
-  def typeCheck() = {
-    rdd.forall { case (k, v) => keySignature.typeCheck(k) && valueSignature.typeCheck(v) } &&
-      KeyTable.toSingleRDD(rdd, nKeys, nValues).forall { a =>
-        signature.typeCheck(a)
-      }
-  }
+  def typeCheck() = rdd.forall { a => signature.typeCheck(a) }
 
   def printSchema(): Unit = println(signature.toPrettyString())
 
+  def withKeys(): RDD[(Annotation, Annotation)] = {
+    val keyIndices = keyFields.map(_.index)
+    val nFieldsLocal = nFields
+
+    rdd.map { a =>
+      val r = KeyTable.annotationToSeq(a, nFieldsLocal).zipWithIndex
+      val keyRow = keyIndices.map(i => r(i)._1)
```

Comment:
Indexing into a Seq is a potentially O(N) operation.  I think we should probably just use arrays everywhere instead of Seqs.  a row `toSeq` operation does produce a `WrappedArray` like we want, but Seq.fill(n)(null) could use something bad, like List or Vector